rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,850 @@
+import cats.effect.Sync
+import cats.syntax.all._
+import coop.rchain.shared.syntax._
+import coop.rchain.store.KeyValueTypedStore
+import scodec.bits.ByteVector
+
+import scala.annotation.tailrec
+import scala.collection.concurrent.TrieMap
+
+object RadixTree {
+
+  /**
+    * Described items which contain in [[Node]].
+    */
+  sealed trait Item
+
+  /**
+    * Empty item
+    */
+  final case object EmptyItem extends Item
+
+  /**
+    * Item which contain data.
+    */
+  final case class Leaf(prefix: ByteVector, value: ByteVector) extends Item
+
+  /**
+    * Item which contain pointer for child [[Node]]
+    */
+  final case class NodePtr(prefix: ByteVector, ptr: ByteVector) extends Item
+
+  /**
+    * Base type for nodes in Radix History.
+    *
+    * Always contains 256 [[Item]]
+    */
+  type Node = Vector[Item]
+
+  /**
+    * Number of items always constant.
+    */
+  val numItems = 256
+
+  /**
+    * Empty node consists only of [[EmptyItem]]'s.
+    */
+  val emptyNode: Node = (0 until numItems).map(_ => EmptyItem).toVector
+
+  /**
+    * Binary codecs for serializing/deserializing Node in Radix tree
+    *
+    * {{{
+    * Coding structure for items:
+    *   EmptyItem                   - Empty (not encode)
+
+    *   Leaf(prefix,value)    -> [item number] [second byte] [prefix0]..[prefixM] [value0]..[value31]
+    *                               where is: [second byte] -> bit7 = 0 (Leaf identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    *   NodePtr(prefix,ptr)   -> [item number] [second byte] [prefix0]..[prefixM] [ptr0]..[ptr31]
+    *                               where is: [second byte] -> bit7 = 1 (NodePtr identifier)
+    *                                                          bit6..bit0 - prefix length = M (from 0 to 127)
+    *
+    * For example encode this Node which contains 2 non-empty items (number 1 and number 2):
+    * (0)[Empty] (1)[Leaf(prefix:0xFFFF,value:0x00..0001)] (2)[NodePtr(prefix:empty,value:0xFF..FFFF)] (3)...(255)[Empty].
+    * Encoded data = 0x0102FFFF0000000000000000000000000000000000000000000000000000000000000001
+    *                  0280FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
+    * where: item 1 (number_secondByte_prefix_value) = 01_02_FFFF_00..0001
+    *        item 2 (number_secondByte_prefix_value) = 02_80_empty_FF..FFFF
+    * }}}
+    */
+  object Codecs {
+    private val defSize  = 32 // Default size for non-empty item data
+    private val headSize = 2  // 2 bytes: first - item number, second - second byte
+
+    /** Serialization [[Node]] to [[ByteVector]]
+      */
+    def encode(node: Node): ByteVector = {
+
+      // Calculating size of serialized data
+      val sizeNode = node.size
+      @tailrec
+      def loopSize(index: Int, size: Int): Int =
+        if (sizeNode > index)
+          node(index) match {
+            case EmptyItem => loopSize(index + 1, size)
```

Comment:
I don't want to do this, because, for these methods to work, I would have to copy byte sequences into intermediate arrays. This may introduce additional delays.